### PR TITLE
lnwallet: require fee reserve routing node inc. anchor channels

### PR DESCRIPTION
**Pull request by cfromknecht**
This commit restricts the funding case for incoming anchor channels when
a node has no on-chain fee reserve such that this may only happen when
the node is not configured for forwarding htlcs.

This is an attempt to disallow a potentially dangerous case where a node
has no utxos available to sweep a disputed HTLC on an upstream channel
with the preimage. Failure to do so allows the upstream node to timeout
the HTLC and take their funds back, resulting in loss of funds for the
intermediary.

We permit this to occur when the node *is not* configured for routing
because the situation can't result in a loss of funds scenario. The
worst case is that the HTLC remains on chain, and can be timed out
indefinitely (e.g. by adding some funds to the node at a later point int
time) since there is not upstream deadline to meet.

### CODE REVIEW

File: lnwallet/wallet.go [hunk 1]
```diff
@@ -579,8 +579,16 @@ func (l *LightningWallet) PsbtFundingVerify(pendingChanID [32]byte,
...
+	// We also increment the counter when _accepting_ an incoming anchor
+	// channel and the node is configured for HTLC forwarding. This ensures
+	// that we attempt to maintain an onchain UTXO for fee bumping in the
+	// event of unilateral closure. Not maintain a fee bumping UTXO is
+	// dangerous when forwarding payments, since the node may not be able to
+	// sweep a disputed HTLC on the upstream channel with the preimage,
+	// resulting in loss of funds.
 	if pendingReservation.partialState.ChanType.HasAnchors() &&
-		intent.LocalFundingAmt() > 0 {
+		(intent.LocalFundingAmt() > 0 || l.Cfg.IsRoutingNode) {
```
Comment (halseth) [inline lnwallet/wallet.go:591]: this check must also be done in the wallet-funding scenario elsewhere in this file

File: lnwallet/wallet.go [hunk 2]
```diff
@@ -579,8 +579,16 @@ func (l *LightningWallet) PsbtFundingVerify(pendingChanID [32]byte,
 	// but only if we are contributing funds to the channel. This is done
 	// to still allow incoming channels even though we have no UTXOs
 	// available, as in bootstrapping phases.
+	//
+	// We also increment the counter when _accepting_ an incoming anchor
+	// channel and the node is configured for HTLC forwarding. This ensures
+	// that we attempt to maintain an onchain UTXO for fee bumping in the
+	// event of unilateral closure. Not maintain a fee bumping UTXO is
```
Comment (halseth) [inline lnwallet/wallet.go:586]: s/maintain/maintaining

**Review by halseth — CHANGES_REQUESTED** [review]
concecpt ACK

File: lnwallet/wallet.go (hunk 1 above)
Comment (cfromknecht) [reply lnwallet/wallet.go:591]: good catch, i've added a integration test to test the basic funding flow

File: lntest/itest/lnd_test.go [hunk 3]
```diff
@@ -1369,107 +1366,115 @@ test:
...
+		}
+	}
+}
 
-			// Each time, we'll send Carol a new set of coins in
-			// order to fund the channel.
-			ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
-			err = net.SendCoins(ctxt, btcutil.SatoshiPerBitcoin, carol)
-			require.NoError(t.t, err)
+func runBasicFundingFlow(
```
Comment (Roasbeef) [inline lntest/itest/lnd_test.go:1412]: 👍

File: lntest/itest/lnd_test.go [hunk 4]
```diff
@@ -1363,129 +1360,170 @@ func testBasicChannelFunding(net *lntest.NetworkHarness, t *harnessTest) {
...
-				expType := carolCommitType
+	// Each time, we'll send Carol a new set of coins in order to fund the
+	// channel.
+	ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
+	err = net.SendCoins(ctxt, btcutil.SatoshiPerBitcoin, carol)
+	require.NoError(t.t, err)
 
-				switch daveCommitType {
+	daveArgs := daveCommitType.Args()
+	if rejectHtlcReceiver {
```
Comment (Roasbeef) [inline lntest/itest/lnd_test.go:1437]: The test should also use this to determine if a test failure is acceptable. In the case where we don't fund the receiver, but then reject HTLCs for the receiver, and we're doing an anchor commitment funding, then we'd expect the funding flow to fail.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: lntest/itest/lnd_test.go [hunk 5]
```diff
@@ -1371,17 +1373,35 @@ func testBasicChannelFunding(net *lntest.NetworkHarness, t *harnessTest) {
...
-			tests = append(tests, testCase{
-				carolCommitType: carolCommitType,
-				daveCommitType:  daveCommitType,
-			})
+			isAnchorFunding :=
+				carolCommitType == commitTypeAnchors &&
+					daveCommitType == commitTypeAnchors
+
+			if isAnchorFunding {
+				tests = append(tests, testCase{
```
Comment (halseth) [inline lntest/itest/lnd_test.go:1381]: add a comment here

File: lntest/itest/lnd_test.go [hunk 6]
```diff
@@ -9271,6 +9307,13 @@ func testRevokedCloseRetributionAltruistWatchtowerCase(
 		t.Fatalf("unable to send coins to dave: %v", err)
 	}
 
+	// If we're funding an anchor channel, carol will also need an onchain
+	// reserve in order to accept the channel funding.
+	if anchors {
+		err = net.SendCoins(ctxb, btcutil.SatoshiPerBitcoin, carol)
```
Comment (halseth) [inline lntest/itest/lnd_test.go:9313]: can just send coins to carol regardless of channel type.

**Pull request closed without merging**